sdk/core/core-http/src/policies/bearerTokenAuthenticationPolicy.ts

```diff
@@ -83,18 +85,37 @@ export class BearerTokenAuthenticationPolicy extends BaseRequestPolicy {
         spanOptions: webResource.spanOptions
       }
     });
-    webResource.headers.set(Constants.HeaderConstants.AUTHORIZATION, `Bearer ${token}`);
+    if (token) {
+      webResource.headers.set(Constants.HeaderConstants.AUTHORIZATION, `Bearer ${token}`);
+    }
     return this._nextPolicy.sendRequest(webResource);
   }
 
   private async getToken(options: GetTokenOptions): Promise<string | undefined> {
-    // We reset the cached token some before it expires,
+    // We reset the cached token in a time window before it expires,
     // after that point, we retry the refresh of the token only if the token refresher is ready.
+
     let token = this.tokenCache.getCachedToken();
-    if (!token && this.tokenRefresher.isReady()) {
-      token = await this.tokenRefresher.refresh(options);
-      this.tokenCache.setCachedToken(token);
+    let readyToRefresh = !token && this.tokenRefresher.isReady();
+    let refreshPromise: Promise<AccessToken | undefined> | undefined;
+    if (readyToRefresh) {
+      refreshPromise = this.tokenRefresher.refresh(options);
+    }
+
+    // If we have a local copy of the token, we use it.
+    if (this.token) {
+      token = this.token;
     }
+
+    // If we don't have a token, or if the token has expired, we wait for the refresh promise
+    if (!token || token.expiresOnTimestamp < Date.now()) {
```

Comment:
what about this case:

1. `this.tokenCache.getCachedToken()` returns a valid token
2. `this.token` is expired

Since the token from the cache is valid, `readyToRefresh` will be false and `refreshPromise` will be undefined. Then line 107 assigns the expired token over the one we got from the cache.  Now we get here and try to await undefined.

Reply:
This can be solved by the conditional `if (this.token)` some lines above, if we change it to:

```ts
    // We don't use the local copy if the token cache has a token that has an expiration date greater than the local copy.
    if (this.token && this.token.expiresOnTimestamp > (token?.expiresOnTimestamp || 0)) {
      token = this.token;
    }
 ```
 
 I know it's not pretty! But I'll push it while we come up with something better.